### DISCUSSION TITLE
CSharp Compiler Config

### DISCUSSION THREAD

**Discussion by obiwanjacobi** [dotnet/csharplang#3924] [General] [answered]
I want to be able to enforce use and disuse of language features.
Looking at C++ as a warning for what is to come and how it is evolving, you want to be able to move your code base (and team) forward to newer -and hopefully more efficient- ways of doing things.

As a CSharp language you cannot drop existing features. I think we all can understand that. But as a developer I can. Just give me the option to declare those choices. Similar to how the nullable reference types was introduced.

Think of it as an `.editorconfig` but for the compiler.
[2c]

**Comment by YairHalberstadt** [comment, marked as answer]
Analyzers are the solution to this problem.

**Reply by CyrusNajmabadi** [reply]
This the correct answer.  The compiler/platform already provides a way for you to "declare those choices" in an extremely flexible manner.  You can blanket disallow the features, or you can make the determination in terms of absolutely any criteria you care about.  For example, you might be ok if a 3rd party analyzer you use generates these constructs even if you want them written by humans themselves in your own code.

**Reply by obiwanjacobi** [reply]
What Analyzer can I use for this?

**Reply by YairHalberstadt** [reply]
Depending on what you want to forbid you may be able to find an existing analyzer, or you'll have to write one yourself.

**Comment by PathogenDavid** [comment]
> Think of it as an `.editorconfig` but for the compiler.

You can already configure [many C#-specific coding convention preferences via EditorConfig](https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-code-style-settings-reference). Many of them are basically whether or not you want to permit use of older features. So depending on the restrictions you want to impose you might not even need to make an analyzer.

**Comment by spydacarnage** [comment]
I believe the implication is that you can write your own to suit your
specific needs.
